stan/math/prim/meta/promote_scalar_type.hpp

```diff
@@ -0,0 +1,69 @@
+#ifndef STAN_MATH_PRIM_SCAL_META_PROMOTE_SCALAR_TYPE_HPP
+#define STAN_MATH_PRIM_SCAL_META_PROMOTE_SCALAR_TYPE_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Template metaprogram to calculate a type for converting a
+ * convertible type.  This is the base case.
+ *
+ * @tparam T result scalar type.
+ * @tparam S input type
+ */
+template <typename T, typename S, typename Enable = void>
+struct promote_scalar_type {
+  /**
+   * The promoted type.
+   */
+  using type = T;
+};
+
+/**
+ * Template metaprogram to calculate a type for a container whose
+ * underlying scalar is converted from the second template
+ * parameter type to the first.
+ *
+ * @tparam T result scalar type.
+ * @tparam S input type
+ */
+template <typename T, typename S>
+struct promote_scalar_type<T, std::vector<S>> {
+  /**
+   * The promoted type.
+   */
+  using type = std::vector<typename promote_scalar_type<T, S>::type>;
+};
+
+/**
+ * Template metaprogram to calculate a type for a matrix, vector, row vector or
+ * Eigen::Array whose underlying scalar is converted from the second template
+ * parameter type to the first.
+ *
+ * @tparam T result scalar type.
+ * @tparam S input matrix type
+ */
+template <typename T, typename S>
+struct promote_scalar_type<T, S, require_eigen_t<S>> {
+  /**
+   * The promoted type.
+   */
+  using type = typename std::conditional<
+      std::is_same<typename Eigen::internal::traits<std::decay_t<S>>::XprKind,
+                   Eigen::MatrixXpr>::value,
+      Eigen::Matrix<typename promote_scalar_type<T, typename S::Scalar>::type,
+                    S::RowsAtCompileTime, S::ColsAtCompileTime>,
+      Eigen::Array<typename promote_scalar_type<T, typename S::Scalar>::type,
+                   S::RowsAtCompileTime, S::ColsAtCompileTime>>::type;
```

Comment:
Trying to make sure I'm reading this correctly.

"If `S` is an Eigen expression type is an Eigen matrix whose `Scalar` is the promoted `Scalar` type of T and S. Otherwise type is an Eigen array whose `Scalar` type is the promoted type `Scalar` type of `T` and `S`"

Can you clarify what this is for? In isolation I'm not getting the logic

Reply:
For example if we have some container that could be either eigen type, sdt::vector ... and we want the same container type with a different scalar we can use this. I needed this in changes to `empty_broadcast_array`.

Again I made no changes to this, just moved it to /meta.